### ISSUE TITLE
Pytorch Implementation. The validation loss seems to be oscillating whereas the training loss converges

### ISSUE THREAD

**Issue by AshishRMenon** [animgoeth/ARA-CNN#10]
**This is my model definition and structure:**

def init_kernel(m):
    
    if isinstance(m, nn.Conv2d):
        print(m)
        # Initialize kernels of Conv2d layers as kaiming normal
        nn.init.kaiming_normal_(m.weight, mode='fan_in', nonlinearity='relu')
        # Initialize biases of Conv2d layers at 0
        nn.init.zeros_(m.bias)
    
    if isinstance(m, nn.Linear):
        print(m)
        # Initialize kernels of Conv2d layers as kaiming normal
        nn.init.xavier_uniform_(m.weight,gain=nn.init.calculate_gain('relu'))
        # Initialize biases of Conv2d layers at 0
        nn.init.zeros_(m.bias)

class model_ara(nn.Module):
  
    def __init__(self):
        super(model_ara, self).__init__()
        self.conv = nn.Sequential(nn.Conv2d(3,64,kernel_size=7,stride=4,padding=2),nn.BatchNorm2d(64),nn.LeakyReLU(0.1),nn.MaxPool2d(2))
        self.res = nn.Sequential(nn.Conv2d(64,64,kernel_size=3,stride=1,padding=1),nn.BatchNorm2d(64),nn.LeakyReLU(0.1))
        self.avgpool = nn.AvgPool2d(2)
        self.dropout_layer = nn.Sequential(nn.Linear(64,32),nn.LeakyReLU(0.1),nn.Dropout(0.5),nn.Linear(32,8))
        self.softmax = nn.Softmax()
        self.conv.apply(init_kernel)
        self.res.apply(init_kernel)
        self.dropout_layer.apply(init_kernel)
        
        for m in self.modules():
            if isinstance(m, nn.BatchNorm2d):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)    

    def forward(self,input):
        x = self.conv(input) 
        #--------------------------------------------------#

        #------------Residual Block1------------------------#
        for i in range(4):
            x1 = self.res(x)
            x = x+x1

        x = self.avgpool(x)
        
        
        #--------Aux_output (dropout1)--------------------------#
        x_aux = F.adaptive_avg_pool2d(x, (1, 1))
        x_aux = x_aux.view(-1,x_aux.size(1)*x_aux.size(2)*x_aux.size(3))
        x_aux = self.dropout_layer(x_aux)

        

        #-----------Residual Block2-----------------------#
        for i in range(3):
            x1 = self.res(x)
            x = x+x1

        x = self.avgpool(x)

        
        #--------Actual_output--------------#
        x_main = F.adaptive_avg_pool2d(x, (1, 1))
        x_main = x_main.view(-1,x_main.size(1)*x_main.size(2)*x_main.size(3))
        x_main = self.dropout_layer(x_main)
        return x_aux,x_main 

def init_kernel(m):
    
    if isinstance(m, nn.Conv2d):
        print(m)
        # Initialize kernels of Conv2d layers as kaiming normal
        nn.init.kaiming_normal_(m.weight, mode='fan_in', nonlinearity='relu')
        # Initialize biases of Conv2d layers at 0
        nn.init.zeros_(m.bias)
    
    if isinstance(m, nn.Linear):
        print(m)
        # Initialize kernels of Conv2d layers as kaiming normal
        nn.init.xavier_uniform_(m.weight,gain=nn.init.calculate_gain('relu'))
        # Initialize biases of Conv2d layers at 0
        nn.init.zeros_(m.bias)

class model_ara(nn.Module):
  
    def __init__(self):
        super(model_ara, self).__init__()
        self.conv = nn.Sequential(nn.Conv2d(3,64,kernel_size=7,stride=4,padding=2),nn.BatchNorm2d(64),nn.LeakyReLU(0.1),nn.MaxPool2d(2))
        self.res = nn.Sequential(nn.Conv2d(64,64,kernel_size=3,stride=1,padding=1),nn.BatchNorm2d(64),nn.LeakyReLU(0.1))
        self.avgpool = nn.AvgPool2d(2)
        self.dropout_layer = nn.Sequential(nn.Linear(64,32),nn.LeakyReLU(0.1),nn.Dropout(0.5),nn.Linear(32,8))
        self.softmax = nn.Softmax()
        self.conv.apply(init_kernel)
        self.res.apply(init_kernel)
        self.dropout_layer.apply(init_kernel)
        
        for m in self.modules():
            if isinstance(m, nn.BatchNorm2d):
                nn.init.constant_(m.weight, 1)
                nn.init.constant_(m.bias, 0)    

    def forward(self,input):
        x = self.conv(input) 
        #--------------------------------------------------#

        #------------Residual Block1------------------------#
        for i in range(4):
            x1 = self.res(x)
            x = x+x1

        x = self.avgpool(x)
        
        
        #--------Aux_output (dropout1)--------------------------#
        x_aux = F.adaptive_avg_pool2d(x, (1, 1))
        x_aux = x_aux.view(-1,x_aux.size(1)*x_aux.size(2)*x_aux.size(3))
        x_aux = self.dropout_layer(x_aux)

        

        #-----------Residual Block2-----------------------#
        for i in range(3):
            x1 = self.res(x)
            x = x+x1

        x = self.avgpool(x)

        
        #--------Actual_output--------------#
        x_main = F.adaptive_avg_pool2d(x, (1, 1))
        x_main = x_main.view(-1,x_main.size(1)*x_main.size(2)*x_main.size(3))
        x_main = self.dropout_layer(x_main)
        return x_aux,x_main 

**L2 Regularization for the specified layers**

def compute_l2_reg(model):
    # Function that sets weight_decay only for weights and not biases and only 
    lambda_ = 0.0001
    params_dict = dict(model_ft.named_parameters())
    l2_reg=[]
    for key, value in params_dict.items():
        if (key == 'conv.0.weight' or key == 'res.0.weight' or key == 'dropout_layer.0.weight' ):
            l2_reg += [lambda_*torch.norm(value.view(value.size(0),-1),2)]

    l2_reg = sum(l2_reg)
    return l2_reg

**Training Loop**
def train_model(model, criterion, optimizer, scheduler_train , scheduler_val, num_epochs=10):
    since=time.time()
    best_model_wts=copy.deepcopy(model.state_dict())
    best_acc = 0.0
    thresh = 1000
    for epoch in range(num_epochs):
        print('-'*10 + ' Epoch {}'.format(epoch) + '-'*10, flush=True)
            
        for phase in ['train','val']:
            print(phase,flush=True)
            if phase=='train':
                model.train()
            else:
                model.eval()
                
            running_loss_aux=0
            running_loss_out=0
            running_corrects_aux=0
            running_corrects_out=0
            for images , labels in dataloaders[phase]:
                images=images.to(device)
                labels=labels.to(device)
                optimizer.zero_grad()        
                with torch.set_grad_enabled(phase=='train'):
    
                    outputs=model(images)                    
                    _,predictions_main = torch.max(outputs[1],1)
                    _,predictions_aux = torch.max(outputs[0],1)
                    
                    
                    loss_out=criterion(outputs[1],labels)
                    loss_aux=criterion(outputs[0],labels)
                    loss = 0.9*loss_out+0.1*loss_aux + compute_l2_reg(model) #+ l2_conv1 + l2_conv2  + l2_lin1 #+l2_lin2
                    
                    if phase=='train':
                        loss.backward()
                        optimizer.step()
                running_loss_aux+=loss_aux.item()*images.size(0)
                running_loss_out+=loss_out.item()*images.size(0)
#                 predictions_out_binarized = (predictions_out == 0) * 1
#                 labels_binarized = (labels== 0) * 1
#                 predictions_aux_binaized = (predictions_aux == 0) * 1
                running_corrects_out+=torch.sum(predictions_main == labels.data)
                running_corrects_aux+=torch.sum(predictions_aux == labels.data)
                
            epoch_loss_out=running_loss_out/datalength[phase]
            epoch_loss_aux=running_loss_aux/datalength[phase]
            epoch_acc_out=running_corrects_out.double()/datalength[phase]
            epoch_acc_aux=running_corrects_aux.double()/datalength[phase]
            
            
            
            
            if phase=='train':
                scheduler_train.step(epoch_loss_out)
            
            if phase=='val':
                scheduler_val.step(epoch_acc_out)
                acc_list_val.append(epoch_acc_out)
                loss_list_val.append(epoch_loss_out)
            
            
            epoch_list.append(epoch)
            
            print('{} Loss_aux: {:.4f} Loss_out: {:.4f} Acc_aux: {:.4f} Acc_out: {:.4f} '.format(
                phase, epoch_loss_aux, epoch_loss_out ,epoch_acc_aux,epoch_acc_out),flush=True) 

            if phase == 'val' and epoch_acc_out>best_acc:
                best_acc=epoch_acc_out
                best_model_wts=copy.deepcopy(model.state_dict())
                torch.save(model.state_dict(),'/home/ashishmenon/ARA/Pytorch_imp/trained_acc.pth')
            
            if phase == 'val' and epoch_loss_out<thresh:
                thresh = epoch_loss_out
                torch.save(model.state_dict(),'/home/ashishmenon/ARA/Pytorch_imp/trained_loss.pth')
            
            
    model.load_state_dict(best_model_wts)   
    

    return model

model_ft = model_ara().to(device)
criterion = nn.CrossEntropyLoss()
optimizer_ft = optim.Adam(model_ft.parameters(),lr=0.07)
exp_lr_scheduler_val = optim.lr_scheduler.ReduceLROnPlateau(optimizer_ft,verbose=True,mode='max',patience=10)
exp_lr_scheduler_train = optim.lr_scheduler.ReduceLROnPlateau(optimizer_ft,verbose=True,mode='min',patience=10)

model_best = train_model(model_ft, criterion, optimizer_ft,exp_lr_scheduler_train,exp_lr_scheduler_val, num_epochs=100)
